ntoskrnl/ps/job.c

```diff
@@ -141,66 +315,51 @@ NtAssignProcessToJobObject (
     I open the process handle before the job handle is that a simple test showed
     that it first complains about a invalid process handle! The other way around
     would be simpler though... */
-    Status = ObReferenceObjectByHandle(
-        ProcessHandle,
-        PROCESS_TERMINATE,
-        PsProcessType,
-        PreviousMode,
-        (PVOID*)&Process,
-        NULL);
-    if(NT_SUCCESS(Status))
+    Status = ObReferenceObjectByHandle(ProcessHandle,
+                                       PROCESS_TERMINATE,
+                                       PsProcessType,
+                                       PreviousMode,
+                                       (PVOID*)&Process,
+                                       NULL);
+    if (NT_SUCCESS(Status))
     {
-        if(Process->Job == NULL)
+        PEJOB Job;
+        Status = ObReferenceObjectByHandle(JobHandle,
+                                           JOB_OBJECT_ASSIGN_PROCESS,
+                                           PsJobType,
+                                           PreviousMode,
+                                           (PVOID*)&Job,
+                                           NULL);
+        if (NT_SUCCESS(Status))
         {
-            PEJOB Job;
-
-            Status = ObReferenceObjectByHandle(
-                JobHandle,
-                JOB_OBJECT_ASSIGN_PROCESS,
-                PsJobType,
-                PreviousMode,
-                (PVOID*)&Job,
-                NULL);
-            if(NT_SUCCESS(Status))
+            if (ExAcquireRundownProtection(&Process->RundownProtect))
             {
-                /* lock the process so we can safely assign the process. Note that in the
-                meanwhile another thread could have assigned this process to a job! */
-
-                if(ExAcquireRundownProtection(&Process->RundownProtect))
+                if (Process->Job == NULL && PsGetProcessSessionId(Process) == Job->SessionId)
                 {
-                    if(Process->Job == NULL && PsGetProcessSessionId(Process) == Job->SessionId)
-                    {
-                        /* Just store the pointer to the job object in the process, we'll
-                        assign it later. The reason we can't do this here is that locking
-                        the job object might require it to wait, which is a bad thing
```

Comment:
This old comment seems to indicate you shouldn't call PspAssignProcessToJob while holding a rundown count?

Reply:
I am not conviced that I handle rundown correctly at all